File: go-pr/48/go/code-quality-error-handling/string_parser.go
package main

import (
	"fmt"
	"strconv"
)

type StringParser struct{}

func (sp *StringParser) ParseInteger(str string) int {
	// {fact rule=code-quality-error-handling@v1.0 defects=1}
	num, _ := strconv.Atoi(str)
	return num
	// {/fact}
}

func (sp *StringParser) SafeParseInteger(str string) (int, error) {
	// {fact rule=code-quality-error-handling@v1.0 defects=0}
	if str == "" {
		return 0, fmt.Errorf("empty string cannot be parsed")
	}
	
	num, err := strconv.Atoi(str)
	if err != nil {
		return 0, fmt.Errorf("failed to parse '%s' as integer: %w", str, err)
	}
	return num, nil
	// {/fact}
}

func (sp *StringParser) ParseFloat(str string) float64 {
	// {fact rule=code-quality-error-handling@v1.0 defects=1}
	num, _ := strconv.ParseFloat(str, 64)
	return num
	// {/fact}
}

func (sp *StringParser) SafeParseFloat(str string) (float64, error) {
	// {fact rule=code-quality-error-handling@v1.0 defects=0}
	if str == "" {
		return 0, fmt.Errorf("empty string cannot be parsed")
	}
	
	num, err := strconv.ParseFloat(str, 64)
	if err != nil {
		return 0, fmt.Errorf("failed to parse '%s' as float: %w", str, err)
	}
	return num, nil
	// {/fact}
}

func main() {
	parser := &StringParser{}
	num := parser.ParseInteger("123")
	fmt.Printf("Number: %d\n", num)
}

File: ts-pr/55/typescript/code-quality-error-handling/string_parser.ts
class StringParser {
    parseInteger(str: string): number {
        // {fact rule=code-quality-error-handling@v1.0 defects=1}
        return parseInt(str);
        // {/fact}
    }

    safeParseInteger(str: string): number {
        // {fact rule=code-quality-error-handling@v1.0 defects=0}
        if (!str || str.trim() === '') {
            throw new Error('Empty string cannot be parsed');
        }
        
        const result = parseInt(str);
        if (isNaN(result)) {
            throw new Error(`Failed to parse '${str}' as integer`);
        }
        return result;
        // {/fact}
    }

    parseFloat(str: string): number {
        // {fact rule=code-quality-error-handling@v1.0 defects=1}
        return parseFloat(str);
        // {/fact}
    }

    safeParseFloat(str: string): number {
        // {fact rule=code-quality-error-handling@v1.0 defects=0}
        if (!str || str.trim() === '') {
            throw new Error('Empty string cannot be parsed');
        }
        
        const result = parseFloat(str);
        if (isNaN(result)) {
            throw new Error(`Failed to parse '${str}' as float`);
        }
        return result;
        // {/fact}
    }

    parseJSON<T>(jsonStr: string): T {
        // {fact rule=code-quality-error-handling@v1.0 defects=1}
        return JSON.parse(jsonStr);
        // {/fact}
    }

    safeParseJSON<T>(jsonStr: string): T {
        // {fact rule=code-quality-error-handling@v1.0 defects=0}
        if (!jsonStr || jsonStr.trim() === '') {
            throw new Error('Empty JSON string cannot be parsed');
        }
        
        try {
            return JSON.parse(jsonStr);
        } catch (error) {
            throw new Error(`Failed to parse JSON: ${error}`);
        }
        // {/fact}
    }
}

const parser = new StringParser();
const num = parser.parseInteger('123');
console.log(`Number: ${num}`);